### PR TITLE
Execute multi-node requests using try_request.

### PR DESCRIPTION
**Pull request by nihohit**
This means multi node operations now get error handling, and can trigger MOVED, etc.

Additionally, `func` was removed from `CmdArg::Cmd/Pipeline`. This removes the need 
to allocate a box on every request, and reduces code complexity by eliminating an
unnecessary generic type.

### CODE REVIEW

File: redis/src/cluster_async/mod.rs [hunk 1]
```diff
@@ -699,20 +711,30 @@ where
 
     async fn try_cmd_request(
```
Comment (barshaul) [inline redis/src/cluster_async/mod.rs:712]: This function is a bit hard to understand. can you add some comments, e.g., in the first part - single node request without routing, second - multi node request, third - single node request with routing. something that will explain it a bit more, I got lost between the CommandRouting, routing, route_option, RoutingInfo...

File: redis/src/cluster_async/mod.rs [hunk 2]
```diff
@@ -454,12 +449,12 @@ where
         let inner = Arc::new(InnerCore {
             conn_lock: RwLock::new((connections, Default::default())),
             cluster_params,
+            pending_requests: Mutex::new(Vec::new()),
```
Comment (barshaul) [inline redis/src/cluster_async/mod.rs:452]: why did you move it to InnerCore?

File: redis/src/cluster_async/mod.rs [hunk 3]
```diff
@@ -611,40 +605,61 @@ where
...
+            res.map_err(|_| RedisError::from((ErrorKind::ResponseError, "request wasn't handled")))
+                .and_then(|res| res.map(extract_result))
+        };
+
+        let get_receiver = |(_, receiver): (_, oneshot::Receiver<RedisResult<Response>>)| async {
+            convert_result(receiver.await)
         };
 
         // TODO - once Value::Error will be merged, these will need to be updated to handle this new value.
         let result = match response_policy {
```
Comment (barshaul) [inline redis/src/cluster_async/mod.rs:646]: maybe move it to a separate function - process_results, or something similar? IMO it's too long

File: redis/src/cluster_async/mod.rs [hunk 4]
```diff
@@ -611,40 +605,61 @@ where
...
             Response::Multiple(_) => unreachable!(),
         };
 
-        let run_func = |(_, conn, cmd)| {
-            Box::pin(async move {
-                let conn = conn.await;
-                Ok(extract_result(func(conn, cmd).await?))
-            })
+        let convert_result = |res: Result<RedisResult<Response>, _>| {
+            res.map_err(|_| RedisError::from((ErrorKind::ResponseError, "request wasn't handled")))
```
Comment (barshaul) [inline redis/src/cluster_async/mod.rs:637]: What does 'request wasn't handled' mean? when do we get this error?

File: redis/src/cluster_async/mod.rs (hunk 2 above)
Comment (nihohit) [reply redis/src/cluster_async/mod.rs:452]: otherwise it won't be accessible from within `execute_on_multiple_nodes`, and you won't be able to add the new requests to the queue.

File: redis/src/cluster_async/mod.rs [hunk 5]
```diff
@@ -699,20 +711,30 @@ where
...
     ) -> (OperationTarget, RedisResult<Response>) {
+        let routing = match routing {
+            CommandRouting::Route(routing) => routing,
+            CommandRouting::Connection { addr, conn } => {
+                if !asking {
+                    let mut conn = conn.await;
+                    let result = conn.req_packed_command(&cmd).await.map(Response::Single);
+                    return (addr.into(), result);
+                }
+                None
```
[the remark this answers is not in the document]
Comment (nihohit) [reply redis/src/cluster_async/mod.rs:728]: > so, if 'asking', we get a random node? why?

No, if `asking` then we'll get the redirected connection when calling `get_connection`.

> If we already have the connection, why won't we just do something like 

Because then we'd use the old connection instead of getting the redirected connection.

File: redis/src/cluster_async/mod.rs (hunk 1 above)
Comment (nihohit) [reply redis/src/cluster_async/mod.rs:712]: added some comments, please tell me if they are sufficient.

**Comment by nihohit** [comment]
@barshaul round

File: redis/src/cluster_async/mod.rs (hunk 1 above)
Comment (barshaul) [reply redis/src/cluster_async/mod.rs:721]: looks great now, thanks

**Review by barshaul — APPROVED** [review]
